src/testapp/sessions/universe.cpp

```diff
@@ -531,4 +533,402 @@ Session setup_testplanets_draw(
+
+    auto const tgUCore = solarSystemCore.get_pipelines<PlUniCore>();
+    auto const tgUSFrm = solarSystemScnFrame.get_pipelines<PlUniSceneFrame>();
+
+    auto& rUniverse = top_get< Universe >(topData, idUniverse);
+
+
+    constexpr int precision = 10;
+
+    // Create coordinate spaces
+    CoSpaceId const mainSpace = rUniverse.m_coordIds.create();
+    std::vector<CoSpaceId> satSurfaceSpaces(c_planetCount);
+    rUniverse.m_coordIds.create(satSurfaceSpaces.begin(), satSurfaceSpaces.end());
+
+    rUniverse.m_coordCommon.resize(rUniverse.m_coordIds.capacity());
+
+    CoSpaceCommon& rMainSpaceCommon = rUniverse.m_coordCommon[mainSpace];
+    rMainSpaceCommon.m_satCount = c_planetCount;
+    rMainSpaceCommon.m_satCapacity = c_planetCount;
+
+    Session out;
+    OSP_DECLARE_CREATE_DATA_IDS(out, topData, TESTAPP_DATA_SOLAR_SYSTEM_PLANETS);
+    auto& rCoordNBody = top_emplace< osp::KeyedVec<CoSpaceId, CoSpaceNBody> >(topData, idCoordNBody);
+    rCoordNBody.resize(rUniverse.m_coordIds.capacity());
+
+    // Associate each planet satellite with their surface coordinate space
+    for (SatId satId = 0; satId < c_planetCount; ++satId)
+    {
+        CoSpaceId const surfaceSpaceId = satSurfaceSpaces[satId];
+        CoSpaceCommon& rCommon = rUniverse.m_coordCommon[surfaceSpaceId];
+        rCommon.m_parent = mainSpace;
+        rCommon.m_parentSat = satId;
+    }
+
+    // Coordinate space data is a single allocation partitioned to hold positions, velocities, and
+    // rotations.
+    // TODO: Alignment is needed for SIMD (not yet implemented). see Corrade alignedAlloc
+
+    std::size_t bytesUsed = 0;
+
+    // Positions and velocities are arranged as XXXX... YYYY... ZZZZ...
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satPositions[0]);
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satPositions[1]);
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satPositions[2]);
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satVelocities[0]);
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satVelocities[1]);
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satVelocities[2]);
+
+    // Rotations use XYZWXYZWXYZWXYZW...
+    partition(bytesUsed, c_planetCount, rMainSpaceCommon.m_satRotations[0],
+        rMainSpaceCommon.m_satRotations[1],
+        rMainSpaceCommon.m_satRotations[2],
+        rMainSpaceCommon.m_satRotations[3]);
+
+    partition(bytesUsed, c_planetCount, rCoordNBody[mainSpace].mass);
+    partition(bytesUsed, c_planetCount, rCoordNBody[mainSpace].radius);
+    partition(bytesUsed, c_planetCount, rCoordNBody[mainSpace].color);
+
+    // Allocate data for all planets
+    rMainSpaceCommon.m_data = Array<unsigned char>{ Corrade::NoInit, bytesUsed };
+
+    std::size_t nextBody = 0;
+    auto const add_body = [&rMainSpaceCommon, &nextBody, &rCoordNBody, &mainSpace]
+    (
+        Vector3l position,
+        Vector3d velocity,
+        Magnum::Vector4d rotation,
```

Comment:
I'd presume it's less convenient to write the individual components with quaternions? currently a minor issue, but this will make it a little more difficult for non-zero rotations, eg: `add_body(.., Quaternion::rotation(angle, Vector3{1.0f, 0.0f, 0.0f}), ...)`

Quaternions consist of a vector and scalar part (honestly, I don't know enough about how they work). For magnum, they can be accessed with:
`rotation.vector().x(), rotation.vector().y(), rotation.vector().z(), rotation.scalar()`